### PR TITLE
Sprite text printing improvements

### PR DESCRIPTION
**Pull request by hedara90**
## Description

Built on base improvements from @Gudf

Currently contains a lot of code that's only here for testing purposes.

## Media

pokeemerald-51

https://github.com/user-attachments/assets/3b73e7a7-6802-4ff6-8ad9-8bf66759c3ba

## Things to note in the release changelog:
- Sprite printing must be set up with `SetupSpritesForTextPrinting(arrayOfSpriteIds, arrayOfSourceSprites, numSpritesX, numSpritesY)`
- Has migration script to handle changes to `IsTextPrinterActive`

## Discord contact info

hedara

### CODE REVIEW

File: graphics/test_sprite.png
Comment (hedara90) [inline graphics/test_sprite.png:1]: Only here for testing purposes, remove before merging.

File: include/debug.h [hunk 1]
```diff
@@ -9,6 +9,8 @@ const struct Trainer* GetDebugAiTrainer(void);
 void DebugNative_GetAbilityNames(void);
 void DebugNative_Party_SetFriendship(void);
 
+void TestThing();
```
Comment (hedara90) [inline include/debug.h]: Only here for testing purposes, remove before merging.

File: include/even_sprite.h
Comment (hedara90) [inline include/even_sprite.h:1]: Only here for testing purposes, remove before merging.

File: src/debug.c [hunk 2]
```diff
@@ -78,6 +78,8 @@
 #include "fake_rtc.h"
 #include "save.h"
 
+#include "even_sprite.h"
```
Comment (hedara90) [inline src/debug.c]: Only here for testing purposes, remove before merging.

File: src/debug.c [hunk 3]
```diff
@@ -4189,3 +4191,87 @@ void CheckEWRAMCounters(struct ScriptContext *ctx)
     ConvertIntToDecimalStringN(gStringVar1, gFollowerSteps, STR_CONV_MODE_LEFT_ALIGN, 5);
     ConvertIntToDecimalStringN(gStringVar2, gChainFishingDexNavStreak, STR_CONV_MODE_LEFT_ALIGN, 5);
 }
+
```
Comment (hedara90) [inline src/debug.c]: Only here for testing purposes, remove before merging.

File: src/even_sprite.c
Comment (hedara90) [inline src/even_sprite.c:1]: Only here for testing purposes, remove before merging.

File: src/field_control_avatar.c [hunk 4]
```diff
@@ -224,7 +224,8 @@ int ProcessPlayerFieldInput(struct FieldInput *input)
     if (input->pressedStartButton)
     {
         PlaySE(SE_WIN_OPEN);
-        ShowStartMenu();
+        //ShowStartMenu();
+        TestThing();
```
Comment (hedara90) [inline src/field_control_avatar.c]: Only here for testing purposes, remove before merging.

File: src/text.c [hunk 5]
```diff
@@ -51,7 +52,7 @@ static u32 GetGlyphWidth_ShortNarrow(u16, bool32);
 static u32 GetGlyphWidth_ShortNarrower(u16, bool32);
 
 static EWRAM_DATA struct TextPrinter sTempTextPrinter = {0};
-static EWRAM_DATA struct TextPrinter sTextPrinters[WINDOWS_MAX] = {0};
+static EWRAM_DATA struct TextPrinter sTextPrinters[WINDOWS_MAX + MAX_SPRITES] = {0};
```
Comment (mrgriffin) [inline src/text.c]: We don't need 96 text printers! afaict `sizeof(struct TextPrinter)` is `0x24`/`36`—so an extra 64 of them weighs in at 2.25kB. I think this ought to look a bit more like heap/sprites/tasks/etc where there's a smaller pool dedicated to sprites, and we allocate out of that.

tbh even GF's 32 text printers must be massive overkill. You only need to store a text printer in RAM if it needs to exist over multiple frames, i.e. it's rendering "animated" text. I'd bet that most of the time at most 1 exists, and in rare circumstances maybe it's as high as 2. But obviously that's not something you need to change in this PR.

File: src/sprite.c [hunk 6]
```diff
@@ -1802,3 +1799,360 @@ u32 GetSpanPerImage(u32 shape, u32 size)
...
+    FillSpriteColor(input->spriteId, input->startX, input->startY, input->width, input->height, input->color, input->spriteWidth);
+}
+
+void SpriteFillRectWithColor(u32 spriteId, u32 left, u32 top, u32 width, u32 height, u32 color)
+{
+    //  Check if area spans more than 1 sprite
+    u32 spriteWidth = gOamDimensions[gSprites[spriteId].oam.shape][gSprites[spriteId].oam.size].width;
+    u32 spriteHeight = gOamDimensions[gSprites[spriteId].oam.shape][gSprites[spriteId].oam.size].height;
+
+    color = color | color << 4 | color << 8 | color << 12 | color << 16 | color << 20 | color << 24 | color << 28;
```
Comment (mrgriffin) [inline src/sprite.c]: EDIT: This comment is irrelevant given that we're likely to replace `FillSpriteColor`, but I've left it here because it's a neat trick.

GCC might be able to spot this for itself, but it's faster to express this as:
```suggestion
    color = color * 0x11111111;
```

File: src/battle_interface.c [hunk 7]
```diff
@@ -953,32 +954,34 @@ static void UpdateLvlInHealthbox(u8 healthboxSpriteId, u8 lvl)
...
+    txtPtr = ConvertIntToDecimalStringN(text, currHp, STR_CONV_MODE_RIGHT_ALIGN, HP_MAX_DIGITS);
     *txtPtr++ = CHAR_SLASH;
-    txtPtr = ConvertIntToDecimalStringN(txtPtr, maxHp, STR_CONV_MODE_LEFT_ALIGN, 4);
-    // Print last 6 chars on the right window
-    windowTileData = AddTextPrinterAndCreateWindowOnHealthbox(txtPtr - 6, 0, 5, bgColor, &windowId);
-    HpTextIntoHealthboxObject(objVram + rightTile, windowTileData, 4);
-    RemoveWindowOnHealthbox(windowId);
+    txtPtr = ConvertIntToDecimalStringN(txtPtr, maxHp, STR_CONV_MODE_LEFT_ALIGN, HP_MAX_DIGITS);
+
+    SpriteFillRectWithColor(spriteId, 4, 4, 12, 12, 10);
```
Comment (hedara90) [inline src/battle_interface.c]: Leftover testing code.
Removed in unpushed commit.

File: src/text.c [hunk 8]
```diff
@@ -748,58 +809,147 @@ inline static void GLYPH_COPY(u8 *windowTiles, u32 widthOffset, u32 x0, u32 y0,
...
+    if (textPrinter->printerTemplate.type == SPRITE_TEXT_PRINTER
+     && glyphHeight != gCurGlyph.height)
+    {
+        sprite = &gSprites[gSprites[textPrinter->printerTemplate.spriteId].data[2]];
+        destTiles = (void*)(OBJ_VRAM0) + sprite->oam.tileNum * TILE_SIZE_4BPP;
+
+        u32 newHeight = gCurGlyph.height - glyphHeight;
+
+        u32 leftHalf[16];
+        u32 rightHalf[16];
```
Comment (hedara90) [inline src/text.c:898]: Add a check for if the right half needs to be handled at all, it's way cheaper in most cases to not do the shift for the right half of the glyph.

File: src/sprite.c [hunk 9]
```diff
@@ -1802,3 +1799,360 @@ u32 GetSpanPerImage(u32 shape, u32 size)
...
+
+    return AddTextPrinter(&printer, 0, NULL);
+}
+
+#define firstSprite data[0]
+#define nextX data[1]
+#define nextY data[2]
+#define firstInRow data[3]
+
+inline static void GLYPH_COPY(u8 *windowTiles, u32 widthOffset, u32 x0, u32 y0, u32 *glyphPixels, s32 width, s32 height)
```
Comment (mrgriffin) [inline src/sprite.c]: Mentioned on [Discord](https://discord.com/channels/419213663107416084/1442814687648550952/1442858669128548515) `GLYPH_COPY` is more general than it needs to be for the purpose of filling a sprite, and that causes it to be less efficient than it should be. In particular there's no need to compute `mask` based on `pixels` (allowing `mask` to be lifted out of the loop), and there's no need to access `alignedWindowTiles[8]` inside the loop (not checked, but I suspect upshot is it GCC to generate a `stmia`).

File: src/text.c (hunk 5 above)
Comment (hedara90) [reply src/text.c]: It'll probably have to be in scope for this PR.
In vanilla, text printers are "allocated" by directly referencing which window they're attached to.
```C
sTextPrinters[printerTemplate->windowId] = sTempTextPrinter;
```
So to reduce the total number of printers, some kind of allocation function is needed.

Comment (hedara90) [reply src/text.c]: I have a change that seems to work, and yea, so far I haven't gotten it to use more than 1 text printer at once, so we could save 30 · 44 bytes of EWRAM.

Comment (hedara90) [reply src/text.c]: Resolved with a8126f2206f0e9d683f5b3ccd9f4436966741b3c

File: src/sprite.c (hunk 9 above)
Comment (hedara90) [reply src/sprite.c]: Replaced with custom fill functions

**Comment by hedara90** [comment]
Only missing gamepad icon handling now.

**Comment by hedara90** [comment]
Ready for review

**Comment by hedara90** [comment]
Follow-up from discussion on discord.
The text printers can be refactored to use heap allocations and a linked list instead of a fixed `EWRAM` allocation.
Implementation currently being explored.

**Comment by hedara90** [comment]
Text printer allocation updated to use a linked list instead of static allocations.
This also means that they're now fully decoupled from window IDs

**Comment by hedara90** [comment]
Not quite ready for review actually.
A more generic `AddSpriteTextPrinter` needs to be added.
edit: They have now been added

File: include/text.h [hunk 10]
```diff
@@ -168,14 +176,18 @@ extern struct TextGlyph gCurGlyph;
...
 void DeactivateAllTextPrinters(void);
 u16 AddTextPrinterParameterized(u8 windowId, u8 fontId, const u8 *str, u8 x, u8 y, u8 speed, void (*callback)(struct TextPrinterTemplate *, u16));
+u16 AddSpriteTextPrinterParametrerized(u8 spriteId, u8 fontId, const u8 *str, u8 x, u8 y, u8 speed, void (*callback)(struct TextPrinterTemplate *, u16));
+void AddSpriteTextPrinterParameterized3(u8 spriteId, u8 fontId, u8 left, u8 top, const u8 *color, s8 speed, const u8 *str);
+void AddSpriteTextPrinterParameterized4(u8 spriteId, u8 fontId, u8 left, u8 top, u8 letterSpacing, u8 lineSpacing, const u8 *color, s8 speed, const u8 *str);
+void AddSpriteTextPrinterParameterized6(u8 spriteId, u8 fontId, u8 left, u8 top, u8 letterSpacing, u8 lineSpacing, const union TextColor color, s8 speed, const u8 *str);
 bool32 AddTextPrinter(struct TextPrinterTemplate *printerTemplate, u8 speed, void (*callback)(struct TextPrinterTemplate *, u16));
 void RunTextPrinters(void);
-bool32 IsTextPrinterActive(u8 id);
+bool32 IsTextPrinterActive(u32 id, bool32);
```
Comment (FosterProgramming) [inline include/text.h]: Personally I would prefer to have two functions one for sprite , one for window

File: src/text.c [hunk 11]
```diff
@@ -50,8 +52,11 @@ static u32 GetGlyphWidth_SmallNarrower(u16, bool32);
 static u32 GetGlyphWidth_ShortNarrow(u16, bool32);
 static u32 GetGlyphWidth_ShortNarrower(u16, bool32);
 
-static EWRAM_DATA struct TextPrinter sTempTextPrinter = {0};
-static EWRAM_DATA struct TextPrinter sTextPrinters[WINDOWS_MAX] = {0};
```
Comment (FosterProgramming) [inline src/text.c:54]: We should probably increase HEAP_SIZE to compensate the printers not being allocated directly anymore (maybe not to the entire size of sTextPrinters[WINDOWS_MAX])

File: src/text.c [hunk 12]
```diff
@@ -630,55 +811,205 @@ inline static void GLYPH_COPY(u8 *windowTiles, u32 widthOffset, u32 x0, u32 y0,
     }
 }
 
-void CopyGlyphToWindow(struct TextPrinter *textPrinter)
+#pragma GCC diagnostic push
```
Comment (FosterProgramming) [inline src/text.c]: What does this do? Is it also a temporary thing?

File: src/text.c [hunk 13]
```diff
@@ -630,55 +811,205 @@ inline static void GLYPH_COPY(u8 *windowTiles, u32 widthOffset, u32 x0, u32 y0,
...
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+u32 CopyGlyphToWindow(struct TextPrinter *textPrinter)
```
Comment (FosterProgramming) [inline src/text.c]: We should change the name if it applies to both sprite and window, it's only called in text.c and braille.c so I don't think it needs to be added to the migration script but mention the name change in the changelog

**Review by FosterProgramming — COMMENTED** [review]
I need to look into spend more time looking at copyGlyph, I feel like there could be possible improvements there but overall it's good. I also don't really having so many parametrized printers but I have no idea on how to improve that at the moment and we can leave that to another PR

File: src/text.c (hunk 12 above)
Comment (hedara90) [reply src/text.c]: It ignores the `-Wmaybe-uninitialized` check until `#GCC diagnostic pop`

File: src/text.c (hunk 11 above)
Comment (hedara90) [reply src/text.c:54]: What would be a reasonable increase then?
36 bytes? or 72?
No vanilla thing uses more than 1 printer at any given time, so I'd assume that 36 would be enough, and if downstream uses more printers they'd have to deal with it, I can add a note about needing to increase heap in that case.

**Comment by hedara90** [comment]
Review comments addressed.

File: src/sprite.c [hunk 14]
```diff
@@ -1798,3 +1795,538 @@ u32 GetSpanPerImage(u32 shape, u32 size)
...
+    u8 startX;
+    u8 startY;
+    u8 width;
+    u8 height;
+    u8 spriteWidth;
+    enum FillMode mode:8;
+    u32 color;
+};
+
+static void FillStructWithColor(struct SpriteToFill *input)
```
Comment (FosterProgramming) [inline src/sprite.c]: Should be refactored, see in DM

**Comment by FosterProgramming** [comment]
I know this is a bad idea and people should probably not do that but is it technically possible to use 8x8 sprites for printing thus meaning a 16x16 glyph could cover up to 9 sprites instead of 4

**Comment by hedara90** [comment]
Review comments addressed.

> I know this is a bad idea and people should probably not do that but is it technically possible to use 8x8 sprites for printing thus meaning a 16x16 glyph could cover up to 9 sprites instead of 4

I have a solution to that...

**Comment by hedara90** [comment]
Updated with handling to forbid extremely small sprites and simplified the sprite fill functions.

**Comment by FosterProgramming** [comment]
Everything is good to me. I will merge once you remove the testing code

**Comment by hedara90** [comment]
Testing code removed

File: graphics/test_sprite.png
Comment (FosterProgramming) [reply graphics/test_sprite.png:1]: Is the test sprite statying?

Comment (hedara90) [reply graphics/test_sprite.png:1]: No, I just missed it

File: src/battle_interface.c [hunk 15]
```diff
@@ -909,32 +909,41 @@ static void UpdateLvlInHealthbox(u8 healthboxSpriteId, u8 lvl)
...
-    u8 *windowTileData;
-    u32 windowId, tilesCount, x;
-    u8 text[28], *txtPtr;
-    void *objVram = (void *)(OBJ_VRAM0) + gSprites[spriteId].oam.tileNum * TILE_SIZE_4BPP;
+    u32 width;
+    u8 text[2 * HP_MAX_DIGITS + 2], *txtPtr;
+
+    const union TextColor color =
+    {
+        .background = 0,
```
Comment (FosterProgramming) [inline src/battle_interface.c:925]: Sorry but one last thing I noticed, the argument bgColor is not used in the function `PrintHpOnHealthbox` anymore so either you remove it from the definition or maybe it's supposed to be here ( like `.background = bgColor`? I'm not sure about that one)

Comment (hedara90) [reply src/battle_interface.c:925]: I need to do some more for the hp printing apparently.

Comment (hedara90) [reply src/battle_interface.c:925]: It should be fixed now

**Review by FosterProgramming — APPROVED** [review]
